in_toto/gpg/common.py

```diff
@@ -127,62 +134,363 @@ def parse_pubkey_payload(data):
+      # - disallow duplicate packets
+      if packet_type != PACKET_TYPE_PRIMARY_KEY and \
+          not key_bundle[PACKET_TYPE_PRIMARY_KEY]["key"]:
+        raise Exception("First packet must be a primary key ('{}'), got '{}'."
+            .format(PACKET_TYPE_PRIMARY_KEY, packet_type))
+
+      elif packet_type == PACKET_TYPE_PRIMARY_KEY and \
+          key_bundle[PACKET_TYPE_PRIMARY_KEY]["key"]:
+        raise Exception("Unexpected primary key.")
+
+      # Fully parse master key to fail early, e.g. if key is malformed
+      # or not supported, but also retain original packet for subkey binding
+      # signature verification
+      elif packet_type == PACKET_TYPE_PRIMARY_KEY:
+        key_bundle[PACKET_TYPE_PRIMARY_KEY] = {
+          "key": parse_pubkey_payload(payload),
+          "packet": packet,
+          "signatures": []
+        }
+
+      # Other non-signature packets in the key bundle include User IDs and User
+      # Attributes, required to verify primary key certificates, and subkey
+      # packets. For each packet we create a new ordered dictionary entry. We
+      # use a dictionary to aggregate signatures by packet below,
+      # and it must be ordered because each signature packet belongs to the
+      # most recently parsed packet of a type.
+      elif packet_type in [PACKET_TYPE_USER_ID, PACKET_TYPE_USER_ATTR,
+          PACKET_TYPE_SUB_KEY]:
+        key_bundle[packet_type][packet] = {
+          "body_len": body_len,
+          "signatures": []
+        }
+
+      # The remaining relevant packets are signatures, required to bind subkeys
+      # to the primary key, or to gather additional information about the
+      # primary key, e.g. expiration date.
+      # A signatures corresponds to the most recently parsed packet of a type,
+      # where the type is given by the availability of respective packets.
+      # We test availability and assign accordingly as per the order of packet
+      # types defined in RFC4880 12.1 (bottom-up).
+      elif packet_type == PACKET_TYPE_SIGNATURE:
+        for _type in [PACKET_TYPE_SUB_KEY, PACKET_TYPE_USER_ATTR,
+            PACKET_TYPE_USER_ID]:
+          if key_bundle[_type]:
+            # Add to most recently added packet's signatures of matching type
+            key_bundle[_type][next(reversed(key_bundle[_type]))]\
+                ["signatures"].append(packet)
+            break
+
+        else:
+          # If no packets are available for any of above types (yet), the
+          # signature belongs to the primary key
+          key_bundle[PACKET_TYPE_PRIMARY_KEY]["signatures"].append(packet)
+
+      else:
+        log.info("Ignoring gpg key packet '{}', we only handle packets of "
+            "types '{}' (see RFC4880 4.3. Packet Tags).".format(packet_type,
+            [PACKET_TYPE_PRIMARY_KEY, PACKET_TYPE_USER_ID,
+            PACKET_TYPE_USER_ATTR, PACKET_TYPE_SUB_KEY,
+            PACKET_TYPE_SIGNATURE]))
+
+    except Exception as e:
```

Comment:
There's a bunch of `except Exception`s across this PR. I wonder why the blanket `except` is required, and (for a few places) why it makes sense to `continue` after logging the exception.

Reply:
Agreed! I do have this concern somewhere in my notes and I meant to add it to the PR description's "caveats/todos" section, as this issue applies to other places in the gpg module too, and I plan on addressing it in a separate PR. Thanks for pointing it out!

Reply:
Re `continue`: I considered that question too and decided that in some cases it is okay to continue the overall key export but ignore a self-certificate, if it
  - belongs to a malformed User ID packet (L307),
  - cannot be parsed  (L315),
  - is invalid (L337),

and, ignore a subkey, if it
  - cannot be parsed (L386, L389),
  - has a corresponding binding signature that cannot be parsed (L400),
  - has not exactly one binding signature (L420), or
  - the binding signature is invalid (L430).

But maybe it makes sense too do abort, since these errors are a sign that something is not fully supported or fishy.

Reply:
Will be addressed in https://github.com/in-toto/in-toto/issues/126#issuecomment-473337539.